### PR TITLE
Fix partial/complete data values for incremental responses, especially when combined with `returnPartialData`

### PR DESCRIPTION
**Pull request by jerelmiller**
Apollo Client 4.0-4.2.x is fairly naive in the way `dataState` is reported for incremental responses. Those versions set `dataState` as `streaming` if the request is still in-flight, regardless of whether the data actually satisfies the definition of `streaming`. `streaming` is meant to represent a state where the only holes in `data` are at `@defer` boundaries that aren't streamed in. Prior to this change, this could lead to runtime crashes due to the inaccuracy (see the changeset for an example).

The other issue is that field `read` functions that modified values on individual fields were not applied correctly in intermediate responses returned by the cache. The `read` functions were run correctly, but due to the completeness checks in `QueryInfo`, the values were never applied.

`data` returned in incremental chunks are now properly reported as `partial`, or in some cases `complete` depending on whether it fulfills the requirements of the query and/or `@defer` boundaries. This change also fixes the issue where field `read` function return values were not applied to intermediate chunks.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved `dataState` reporting for incremental `@defer` and `@stream` responses.
  * Results now correctly distinguish between `partial`, `streaming`, and `complete` based on available cached data.
  * Network status continues to indicate when incremental delivery is still in progress.
  * Improved handling of partial cached results, refetches, errors, and cache merges.

* **New Features**
  * Added internal utilities for analyzing GraphQL fields and deferred fragments.

### CODE REVIEW

File: src/react/hooks/__tests__/useSuspenseQuery/streamGraphQL17Alpha9.test.tsx [hunk 1]
```diff
@@ -1553,155 +1559,150 @@ test("can refetch and respond to cache updates after encountering an error in an
...
+          { __typename: "Friend", id: "1", name: "Luke" },
+          { __typename: "Friend", id: "2" },
+          { __typename: "Friend", id: "3" },
+        ],
+      }),
+      dataState: "partial",
+      networkStatus: NetworkStatus.streaming,
+      error: undefined,
+    });
+  }
```
[the remark this answers is not in the document]
Comment (jerelmiller) [reply src/react/hooks/__tests__/useSuspenseQuery/streamGraphQL17Alpha9.test.tsx:1663]: Fixed in [a0918c0](https://github.com/apollographql/apollo-client/pull/13324/commits/a0918c0440fe0c9d5fc42979ab4021133378c335)

File: src/utilities/internal/collectSiblingFields.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      } else {
+        fragmentCollectedFieldsMap = collectSiblingFields(
+          fragment.selectionSet,
+          context,
+          visitedFragments
+        );
+      }
+
+      mergeFieldMaps(collectedFieldsMap, fragmentCollectedFieldsMap);
+    }
```
[the remark this answers is not in the document]
Comment (jerelmiller) [reply src/utilities/internal/collectSiblingFields.ts:89]: Fixed in [852d963](https://github.com/apollographql/apollo-client/pull/13324/commits/852d96388f969ab864e0a7276b8237108eed2f4a)

File: src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts
Comment (jerelmiller) [inline src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts:1]: This test suite is where most of the complex edge case tests were added. It goes to show how naive the previous rules for `dataState` were when it came to incremental responses!

File: src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts [hunk 3]
```diff
@@ -369,3 +376,4918 @@ it.each([["cache-first"], ["no-cache"]] as const)(
...
+      greeting {
+        message
+        ... on Greeting @defer {
+          recipient {
+            name
+            email
+          }
+        }
+      }
+    }
```
Comment (jerelmiller) [inline src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts:614]: This is an example of where the previous behavior was wrong and dangerous. Because we merge partial cache data with streamed chunks, `recipient` was a key present in the response object. From TypeScript's perspective, `streaming` uses the same type as `complete`, so it reports the missing `email` field as a `string` type (even though its not safe to access since its missing). If you used the presence of `recipient` to determine whether the defer chunk had arrived, you could mistakenly perform string methods on `undefined`.

This is now correctly reported as `partial` so that TypeScript reports the response object as a deep partial object.

File: src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts [hunk 4]
```diff
@@ -369,3 +376,4918 @@ it.each([["cache-first"], ["no-cache"]] as const)(
...
+        }
+        ... on Greeting @defer {
+          recipient {
+            name
+            email
+          }
+        }
+      }
+    }
+  `;
```
Comment (jerelmiller) [inline src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts:828]: Another case which wasn't handled at all with 4.x up until this PR. Here we have overlapping deferred and non-deferred fields. The defer boundary isn't partial after the first chunk because we've fulfilled the requirements of the non-deferred fields, so it should be reported as `streaming`.

File: src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts [hunk 5]
```diff
@@ -369,3 +376,4918 @@ it.each([["cache-first"], ["no-cache"]] as const)(
...
+    dataState: "complete",
+    loading: false,
+    networkStatus: NetworkStatus.ready,
+    partial: false,
+  });
+
+  await expect(stream).not.toEmitAnything();
+});
+
+test("does not treat `__typename`-only presence under a `@defer` as the fragment having started", async () => {
```
Comment (jerelmiller) [inline src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts:4562]: Would love more eyes on this one. Should `__typename` only be considered `partial` or `streaming`?

File: src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts [hunk 6]
```diff
@@ -369,3 +376,4918 @@ it.each([["cache-first"], ["no-cache"]] as const)(
...
+        __typename: "Greeting",
+        message: "Hello world",
+        recipient: { __typename: "Person", name: "Cached Alice" },
+      },
+    },
+    dataState: "complete",
+    loading: true,
+    networkStatus: NetworkStatus.streaming,
+    partial: false,
+  });
```
Comment (jerelmiller) [inline src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts:4772]: Another case that is handled much better in this version. Between the initial chunk + partial data in the cache, we have enough data to satisfy the query because there are no more holes in the data. These are now reported as `dataState` `"complete"` since it is safe to access all fields in the response, even though we still have data streaming in from the server.

File: src/core/__tests__/client.watchQuery/streamDefer20220824.test.ts [hunk 7]
```diff
@@ -62,13 +62,13 @@ test("handles streamed scalar lists", async () => {
...
+    data: {
       scalarList: ["apple"],
-    }),
-    dataState: "streaming",
+    },
+    dataState: "complete",
     loading: true,
     networkStatus: NetworkStatus.streaming,
-    partial: true,
+    partial: false,
```
Comment (jerelmiller) [inline src/core/__tests__/client.watchQuery/streamDefer20220824.test.ts:71]: Much of this PR's test changes are the move from `streaming` to `complete` in cases where the data satisfies the selection set. This is especially true with `@stream` where nearly all cases now report as `complete` rather than `streaming` since there are no holes in the data for a `@stream` response.

File: src/react/hooks/__tests__/useSuspenseQuery/streamGraphQL17Alpha9.test.tsx [hunk 8]
```diff
@@ -439,10 +439,14 @@ test('does not suspend streamed queries with partial data in the cache and using
...
-      }),
-      dataState: "streaming",
+      data: {
+        friendList: [
+          { __typename: "Friend", id: "1", name: "Luke" },
+          { __typename: "Friend", id: "2" },
+          { __typename: "Friend", id: "3" },
+        ],
+      },
+      dataState: "partial",
```
Comment (jerelmiller) [inline src/react/hooks/__tests__/useSuspenseQuery/streamGraphQL17Alpha9.test.tsx:449]: This is the other major change in this PR. Partial list items were truncated after the first chunk, even though `returnPartialData: true` was set. This was a mistake and could cause janky UX if list items suddenly disappear in the UI as the list streams in.

The partial list items are now retained and `dataState` properly set to `partial`. Once there are no more partial items, `dataState` changes to `complete` instead of `streaming` since the data fully satisfies the query and it is safe to access all fields.

**Review by DaleSeo — APPROVED** [review]

**Comment by jerelmiller** [comment]
I added a failing test case in [f1764f1](https://github.com/apollographql/apollo-client/pull/13324/commits/f1764f12d1ae43fba8633b8a54ed631c6bb191f0) which will need to be fixed in 4.3 before we release, however I think its worth getting this out in a new alpha now and following up in a separate PR. I opened issue https://github.com/apollographql/apollo-client/issues/13330 to ensure we captured the case. 

~Note: we will also need to test weather `read` functions are called and returned in the result as well as I believe this is also a missing gap in the current solution.~ added tests in #13333 and copied over here in commit [e7c380b](https://github.com/apollographql/apollo-client/pull/13324/commits/e7c380bb7d0f943bbe1840229d309c96c8495489) with a fix applied in [14498bd](https://github.com/apollographql/apollo-client/pull/13324/commits/14498bd1f796f54a5f94bb0357a6dfcf35d8c717).
